src/app/(profile)/(components)/ProfileList/ProfileListItem.tsx

```diff
@@ -16,20 +20,50 @@ export const ProfileListItem: React.FC<ProfileListItemProps> = ({ data }) => {
   const router = useRouter()
   const plausible = usePlausible()
   const searchParams = useSearchParams()
+  const { data: session } = useSession()
+  const { refreshProfilesWithProfileViews } = useProfiles()
+  const { runAsync } = useAsyncAction()
+
+  const visitedProfile = data.profileViews?.find(
+    (view) => view.viewerId === session?.user?.id,
+  )
+
+  const contactedProfile = data.contactRequests?.find(
+    (contact) => contact.senderEmail === session?.user?.email,
+  )
 
   const handleOpenProfile = (event: React.MouseEvent) => {
     event.preventDefault()
+
+    if (!session) {
+      return router.push(AppRoutes.signIn)
+    }
+
     plausible(PlausibleEvents.OpenProfile, {
       props: { username: data.githubUsername },
     })
-    router.push(`${AppRoutes.profile}/${data.githubUsername}`)
+
+    runAsync(async () => {
+      const profileView = await updateOrCreateProfileViewAction(
+        session?.user.id,
+        data.id,
+      )
+
+      if (profileView) {
+        refreshProfilesWithProfileViews(session?.user.id, data.id, profileView)
```

Comment:
**nit:** we just can do it in-memory, and use returned model, because as long as user does not refresh app, he will have current state of profiles in provider state